pyvrp/cpp/PenaltyManager.h

```diff
@@ -129,12 +136,18 @@ class PenaltyManager
     [[nodiscard]] PenaltyBooster getPenaltyBooster();
 };
 
-unsigned int PenaltyManager::loadPenalty(unsigned int load) const
+unsigned int PenaltyManager::loadPenaltyExcess(unsigned int excessLoad) const
 {
-    if (load > vehicleCapacity)
-        return (load - vehicleCapacity) * capacityPenalty;
+    return excessLoad * capacityPenalty;
+}
 
-    return 0;
+unsigned int PenaltyManager::loadPenalty(unsigned int load) const
+{
+    // Branchless for performance: when load > capacity we return the excess 
+    // load penalty; else zero. Note that when load - vehicleCapacity wraps 
+    // around, we return zero because load > vehicleCapacity evaluates as zero
+    // (so there is no issue here due to unsignedness).
+    return (load > vehicleCapacity) * loadPenaltyExcess(load - vehicleCapacity);
 }
```

Comment:
Clang already gets this right, but gcc needed a bit of help getting rid of the conditional here. Somewhat important because this function gets called _a lot_.

(see [here](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:1,endLineNumber:27,positionColumn:1,positionLineNumber:27,selectionStartColumn:1,selectionStartLineNumber:27,startColumn:1,startLineNumber:27),source:'%23include+%3Calgorithm%3E%0A%0Aextern+size_t+vehicleCapacity%3B%0Aextern+size_t+capacityPenalty%3B%0A%0Aunsigned+int+loadPenaltyExcess(unsigned+int+excessLoad)%0A%7B%0A++++return+excessLoad+*+capacityPenalty%3B%0A%7D%0A%0Aunsigned+int+loadPenaltyBranchless(unsigned+int+load)%0A%7B%0A++++//+Branchless+for+performance:+when+load+%3E+capacity+we+return+the+excess%0A++++//+load+penalty%3B+else+zero.+Note+that+when+load+-+vehicleCapacity+wraps%0A++++//+around,+we+return+zero+because+load+%3E+vehicleCapacity+evaluates+as+zero%0A++++//+(so+there+is+no+issue+here+due+to+unsignedness).%0A++++return+(load+%3E+vehicleCapacity)+*+loadPenaltyExcess(load+-+vehicleCapacity)%3B%0A%7D%0A%0Aunsigned+int+loadPenalty(unsigned+int+load)%0A%7B%0A++++if+(load+%3E+vehicleCapacity)%0A++++++++return+(load+-+vehicleCapacity)+*+capacityPenalty%3B%0A%0A++++return+0%3B%0A%7D%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:52.81926876005156,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g122,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:boost,ver:'173')),options:'-std%3Dc%2B%2B17+-O3',selection:(endColumn:10,endLineNumber:5,positionColumn:10,positionLineNumber:5,selectionStartColumn:10,selectionStartLineNumber:5,startColumn:10,startLineNumber:5),source:1),l:'5',n:'0',o:'+x86-64+gcc+12.2+(Editor+%231)',t:'0')),k:47.18073123994844,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4))